myia/abstract/utils.py

```diff
@@ -272,15 +436,54 @@ def abstract_clone(self, x: object, *args):
 #################
 
 
-@overload(bootstrap=True)
-async def abstract_clone_async(self, x: AbstractScalar):
+@overload.wrapper(
+    initial_state=lambda: CloneState({}, None, None)
+)
+async def abstract_clone_async(__call__, self, x, *args):
     """Clone an abstract value (asynchronous)."""
+    async def proceed():
+        if isinstance(x, AbstractValue) and x in cache:
+            return cache[x]
+
+        call = __call__(self, x, *args)
+        if isinstance(call, AsyncGeneratorType):
+            cls = await call.asend(None)
+            inst = cls.empty()
+            cache[x] = inst
+            constructor = _make_constructor(inst)
+            rval = await call.asend(constructor)
+            rval = rval.intern()
+            cache[x] = rval
+            return rval
+        else:
+            return await call
+
+    cache = self.state.cache
+    prop = self.state.prop
+    if prop:
+        if hasattr(x, prop):
+            return getattr(x, prop)
+        elif isinstance(x, (AbstractValue, Context, Reference)):
+            if self.state.check(x, *args):
+                res = x
+            else:
+                res = await proceed()
+            setattr(x, prop, res)
+            return res
+        else:
+            return await proceed()
+    else:
+        return await proceed()
+
+
+@overload  # noqa: F811
+async def abstract_clone_async(self, x: AbstractScalar):
     return AbstractScalar(await self(x.values))
 
 
 @overload  # noqa: F811
 async def abstract_clone_async(self, x: AbstractFunction):
-    return AbstractFunction(*(await self(x.get_sync())))
+    yield (yield AbstractFunction)(*(await self(x.get_sync())))
```

Comment:
I don't get why we need two yields here.  I'm not saying it is wrong, I just don't get it.

Reply:
See my other comment, async generators cannot return values.

Reply:
Ok.